src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollection.cs

```diff
@@ -81,13 +103,117 @@ public int IndexOf(ServiceDescriptor item)
         /// <inheritdoc />
         public void Insert(int index, ServiceDescriptor item)
         {
+            // In theory, we should preserve the relative order here but it's expensive to find that out
+            // and we handle operations that rely on the relative order already
+            AddToServiceTypeMap(item);
+
             _descriptors.Insert(index, item);
         }
 
         /// <inheritdoc />
         public void RemoveAt(int index)
         {
+            // This will throw if the index is out of range
+            ServiceDescriptor item = _descriptors[index];
+            if (_serviceTypes.TryGetValue(item.ServiceType, out List<ServiceDescriptor> items))
+            {
+                items.Remove(item);
+            }
+
             _descriptors.RemoveAt(index);
         }
+
+        internal bool TryAdd(ServiceDescriptor item)
+        {
+            if (!_serviceTypes.TryGetValue(item.ServiceType, out List<ServiceDescriptor> items))
+            {
+                _serviceTypes[item.ServiceType] = new List<ServiceDescriptor> { item };
+                _descriptors.Add(item);
+                return true;
+            }
+            return false;
+        }
+
+        internal bool TryAddEnumerable(ServiceDescriptor item, Type implementationType)
```

Comment:
If we aren't going to make these APIs public, it is kind of pointless to have them return `bool` when no one consumes it.

Reply:
Done